packages/components/src/image-cropper/reducer.ts

```diff
@@ -0,0 +1,286 @@
+/**
+ * Internal dependencies
+ */
+import type { Position, Size } from './types';
+import { rotatePoint, degreeToRadian, getFurthestVector } from './math';
+
+export type State = {
+	// The container/image's width.
+	width: number;
+	// The container/image's height.
+	height: number;
+	// The rotation angle between -45deg to 45deg.
+	angle: number;
+	// The number of 90-degree turns.
+	turns: 0 | 1 | 2 | 3;
+	// The zoom scale.
+	scale: number;
+	// Whether the image is flipped horizontally.
+	flipped: boolean;
+	// The offset position of the cropper window.
+	offset: Position;
+	// The position of center of the image.
+	position: Position;
+	// The size of the cropper window.
+	size: Size;
+	// Whether the cropper window is resizing.
+	isResizing: boolean;
+	// Whether the image is dragging/moving.
+	isDragging: boolean;
+};
+
+type Action =
+	| { type: 'ZOOM'; scale: number }
+	| { type: 'ZOOM_BY'; deltaScale: number }
+	| { type: 'FLIP' }
+	| { type: 'ROTATE'; angle: number }
+	| { type: 'ROTATE_CLOCKWISE'; isCounterClockwise?: boolean }
+	| { type: 'TRANSLATE'; offset: Position }
+	| { type: 'MOVE'; x: number; y: number }
+	| { type: 'RESIZE_START' }
+	| { type: 'RESIZE_WINDOW'; direction: string; delta: Size }
+	| { type: 'RESET' };
+
+function createInitialState( {
+	width,
+	height,
+}: {
+	width: number;
+	height: number;
+} ): State {
+	return {
+		width,
+		height,
+		angle: 0,
+		turns: 0,
+		scale: 1,
+		flipped: false,
+		offset: { x: 0, y: 0 },
+		position: { x: 0, y: 0 },
+		size: { width, height },
+		isResizing: false,
+		isDragging: false,
+	};
+}
+
+function imageCropperReducer( state: State, action: Action ) {
```

Comment:
It'd be great to add some inline comments to the code within this reducer. Geometry code like this is notoriously hard to read.

I spotted a bug, if I resize twice from the bottom-right, the top-left of the resizable box jumps position incorrectly:

https://github.com/user-attachments/assets/aff6fcf0-c210-4e9f-9374-135216cfb29e

I found it a little difficult to understand the intention of the code in the `RESIZE_WINDOW` action, so some comments would help greatly.

Reply:
> I found it a little difficult to understand the intention of the code in the `RESIZE_WINDOW` action, so some comments would help greatly.

Yeah, that's one of the things I need help with 😅. Half the time I don't even know what I'm doing 🤦. I'd need to re-organize my thoughts and refactor this in a way that makes it easier to write comments.

Reply:
The comment here might help - https://github.com/WordPress/gutenberg/pull/63335#discussion_r1680440743.

There might be one or two things I got wrong because I'm mostly doing this as a mental exercise, but I think a good rule is to not modify the underlying image coordinates until the user hits 'apply'. Instead store the manipulations in a 'transform' object that describes how the image has been manipulated. It might be similar to what you're doing already, but I found it difficult to tell.

I think the important thing is to get the reducer working correctly and then have the UI work from that state. It might be challenging with some of the opinionated UI pieces like `resizableBox`, but a solid foundation for the maths will make everything work much more reliably.